crates/accelerate/src/target_transpiler/mod.rs

```diff
@@ -0,0 +1,1285 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+#![allow(clippy::too_many_arguments)]
+
+mod errors;
+mod instruction_properties;
+
+use std::ops::Index;
+
+use ahash::RandomState;
+
+use hashbrown::HashSet;
+use indexmap::{
+    map::{Keys, Values},
+    IndexMap, IndexSet,
+};
+use itertools::Itertools;
+use pyo3::{
+    exceptions::{PyAttributeError, PyIndexError, PyKeyError, PyValueError},
+    prelude::*,
+    pyclass,
+    types::{PyDict, PyList, PySet, PyTuple},
+};
+
+use qiskit_circuit::circuit_instruction::{
+    convert_py_to_operation_type, operation_type_and_data_to_py,
+};
+use qiskit_circuit::operations::{Operation, OperationType, Param};
+use smallvec::{smallvec, SmallVec};
+
+use crate::nlayout::PhysicalQubit;
+
+use errors::TargetKeyError;
+use instruction_properties::InstructionProperties;
+
+use self::exceptions::TranspilerError;
+
+mod exceptions {
+    use pyo3::import_exception_bound;
+    import_exception_bound! {qiskit.exceptions, QiskitError}
+    import_exception_bound! {qiskit.transpiler.exceptions, TranspilerError}
+}
+
+// Custom types
+type Qargs = SmallVec<[PhysicalQubit; 2]>;
+type GateMap = IndexMap<String, PropsMap, RandomState>;
+type PropsMap = IndexMap<Option<Qargs>, Option<InstructionProperties>, RandomState>;
+type GateMapState = Vec<(String, Vec<(Option<Qargs>, Option<InstructionProperties>)>)>;
+
+#[derive(Debug, Clone, FromPyObject)]
+pub enum TargetOperation {
+    Normal(NormalOperation),
+    Variable(VariableOperation),
+}
+
+impl IntoPy<PyObject> for TargetOperation {
+    fn into_py(self, py: Python<'_>) -> PyObject {
+        match self {
+            Self::Normal(normal) => normal.into_py(py),
+            Self::Variable(variable) => variable.into_py(py),
+        }
+    }
+}
+
+impl ToPyObject for TargetOperation {
+    fn to_object(&self, py: Python<'_>) -> PyObject {
+        match self {
+            Self::Normal(normal) => normal.to_object(py),
+            Self::Variable(variable) => variable.to_object(py),
+        }
+    }
+}
+
+impl TargetOperation {
+    fn num_qubits(&self) -> u32 {
+        match &self {
+            Self::Normal(normal) => normal.operation.num_qubits(),
+            Self::Variable(_) => 0,
```

Comment:
So `Variable` is when we provide a class instead of a class instance because it could have a variable amount of qubits (I believe we only do this for `ControlFlowOps`). And so if you call for `num_qubits` it will return a `@property`. We typically wouldn't have a need to call for this on a `VariableOp` as the `Target` operates on them differently, so I could just make this `unreachable!()`.